### PR TITLE
allow user to select a maven module to deploy directly from suggestion list.

### PR DESCRIPTION
**Pull request by wangmingliang-ms**
a preDeployTask is added to build the maven module to generate the final artifact(jar/war).
the final jar/war is deployed to Azure actually.
image

this PR is related to https://github.com/microsoft/vscode-azuretools/pull/925

### CODE REVIEW

**Comment by wangmingliang-ms** [comment]
this PR build failed because it depends on https://github.com/microsoft/vscode-azuretools/pull/925

**Review by nturinski — CHANGES_REQUESTED** [review]
While this definitely achieves allowing a user to select a Maven module to deploy, I feel like this strays away very heavily from our extension's design paradigm.  

For example, while it is possible for users to have multiple .csproj files, we do not support having multiple suggestions for what their default deploy path should be.  This wasn't a limitation, but a design decision.  

In cases where users do have multiple .csproj files or multiple functions projects, we recommend that they use a multi-root workspace so each one can be controlled by their own individual workspace settings.

In the provided screenshot, I see that the folders are structured in such a way that they all contain their own "src" folder so it seems reasonable that they could be all put within their own workspace.
image

They don't explicitly say so, but VS Code's project management page for Java (https://code.visualstudio.com/docs/java/java-project) seems to utilize multi-root workspaces rather than having one folder opened with several snapshots inside.  I deduced this by some of the screenshots that show multiple projects opened in the VS Code explorer.

image


The [Project Manager for Java](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-dependency) also only seems to support one project per workspace, though I suspect that it would support multi-root. I can't confirm this, as I only looked at the codebase briefly, but here you can see they default to the currently opened project https://github.com/microsoft/vscode-java-dependency/blob/27c306b770c23b1eba1f9a7c3e70d2793baced68/src/utility.ts#L11.

I would like for the deployment scenarios to remain somewhat consistent between runtimes.  I understand that there will be necessary differences, but there are a lot of special cases written in this specifically for only Java that impacts all of our users.

I think we can achieve a good experience for users by leveraging the settings that we already provide in the extension.

File: tsconfig.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
         "target": "es6",
         "outDir": "out",
         "lib": [
-            "es6"
+            "es2019"
```
[the remark this answers is not in the document]
Comment (wangmingliang-ms) [reply tsconfig.json]: reverted.

File: src/utils/javaUtils.ts [hunk 2]
```diff
@@ -4,44 +4,50 @@
...
 
 export namespace javaUtils {
     const DEFAULT_PORT: string = '8080';
     const PORT_KEY: string = 'PORT';
+    const JAVA_WEB_CONTAINER_REGEX: RegExp = /^(tomcat|wildfly|jboss)/i;
+    const JAVA_SE_REGEX: RegExp = /^java/i;
 
-    function isJavaWebContainerRuntime(runtime: string | undefined): boolean {
-        return !!runtime && /^(tomcat|wildfly)/i.test(runtime);
+    function isJavaWebContainerRuntime(runtime: SiteConfigResource | undefined): boolean {
```
[the remark this answers is not in the document]
Comment (wangmingliang-ms) [reply src/utils/javaUtils.ts:35]: renamed

**Pull request closed without merging**